src/Agent.Worker/ExecutionContext.cs

```diff
@@ -437,6 +437,14 @@ public void InitializeJob(Pipelines.AgentJobRequestMessage message, Cancellation
                 Container = null;
             }
 
+            if (Container != null)
+            {
+                Container.ImageOSChanged += () => {
```

Comment:
This is probably okay in the worker process since it is short lived, but technically speaking you are leaking memory by hooking up an event handler and never unhooking it (i.e. no -=).
You may want to consider creating a non-anonymous method for the event and unhooking it in some tear down method. (I am assuming that initializeJob has some counterpart at the end).

Reply:
I changed this to a non-anonymous method. Container is owned by the ExecutionContext so would it still be a memory leak? I am guessing the Complete() would be the antithesis of InitializeJob() and could put the tear down there.

Reply:
Unless something has changed with C# event handlers, when you attach an even handler the 2 objects are essentially pointing to each other in a way that disposing both objects doesn't allow them to get garbage collected. The garbage collector still thinks they are being referenced. The only way to fix that is to unhook all the event handlers when you want to dispose of the main objects.

Reply:
If the intention is to keep the event handler hooked up for the duration of the worker process, then this is fine. I just wanted to make sure that we don't start a pattern here in the agent of not unhooking event handlers.

Reply:
I have another PR coming which removes the need for the event handler altogether. It would still be worth understanding your concern here.

Reply:
Here's a [descent explanation](https://michaelscodingspot.com/5-techniques-to-avoid-memory-leaks-by-events-in-c-net-you-should-know/) of the problem. I don't have any worries about this particular handler just the pattern.